### DISCUSSION TITLE
cargo-lambda reqwest & cargos needing ring

### DISCUSSION THREAD

**Discussion by DoIevenLift** [cargo-lambda/cargo-lambda#647] [Q&A] [answered]
Hi, 

Apologies for what might be a silly question. I am trying to get a lambda to build with reqwest as a dependency and am running into the below error:

![image](https://github.com/cargo-lambda/cargo-lambda/assets/78464058/26888887-7c4e-41be-b6fe-013a1d11a17c)

The dependencies in are:

`
[dependencies]
lambda_http = "0.11.1"
tokio = { version = "1.36.0" , features = [ "macros" ]}
aws_lambda_events = { version = "0.15.0", default-features = false, features = ["apigw"] }
serde_json = "1.0.116"
reqwest = { version = "0.12.3", default-features = false, features = ["native-tls-vendored"] }
`

If I remove reqwest this will successfully build. If i change native-tls-vendored to rustls-tls then I'll get **ring** build errors. I have cargo cleaned etc and this hasn't solved the issue. 

I am wondering how people have gotten this to work? I've referred to the docs - https://www.cargo-lambda.info/guide/cross-compiling.html but the suggestions aren't working for me.

Happy to be pointed in the right direction. 

Thanks!

**Comment by miikka** [comment, marked as answer]
Based on your screenshot, the problem is in compiling OpenSSL, not ring. What happens is that Cargo is trying to use OpenSSL from the system (`openssl-sys`) package, but that won't work with cross-compiling. One solution is to [use vendored OpenSSL](https://docs.rs/openssl/0.10.34/openssl/#vendored) by adding the following line to your dependencies:

```toml
openssl = { version = "0.10", features = ["vendored"] }
```

This is essentially the same thing that `native-tls-vendored` ends up doing, and that's also why it stops working when you remove it. If you want to understand why OpenSSL (or ring) gets included in the dependency tree, you can use `cargo tree` to inspect it:

```sh
cargo tree --target aarch64-unknown-linux-gnu 
```

**Reply by DoIevenLift** [reply]
Thanks for taking the time to reply - appreciate it. Thanks for the command to!
